### PR TITLE
large refactor of torch models file

### PR DESCRIPTION
**Pull request by tomsilver**
cleaning up the torch models file. there should be NO functional changes. highlights:

* removed a lot of redundant code, especially the training loops
* added a few new base classes
* removed all occurrences of CFG to promote modularity

### CODE REVIEW

File: src/settings.py [hunk 1]
```diff
@@ -170,7 +170,6 @@ class GlobalSettings:
     side_predicate_learner = "no_learning"
 
     # torch model parameters
-    normalization_scale_clip = 1
```
Comment (ronuchit) [inline src/settings.py:173]: isn't removing this a functional change?

File: src/torch_models.py [hunk 2]
```diff
@@ -1,153 +1,323 @@
...
+        Regressor.__init__(self, seed)
+        nn.Module.__init__(self)  # type: ignore
+        self._max_train_iters = max_train_iters
+        self._clip_gradients = clip_gradients
+        self._clip_value = clip_value
+        self._learning_rate = learning_rate
+
+    @abc.abstractmethod
+    def forward(self, tensor_X: Tensor) -> Tensor:
+        """Pytorch forward method."""
```
Comment (ronuchit) [inline src/torch_models.py:97]: PyTorch?

File: src/torch_models.py [hunk 3]
```diff
@@ -1,153 +1,323 @@
...
+    def _predict(self, x: Array) -> Array:
+        tensor_x = torch.from_numpy(np.array(x, dtype=np.float32))
+        tensor_X = tensor_x.unsqueeze(dim=0)
+        tensor_Y = self(tensor_X)
+        tensor_y = tensor_Y.squeeze(dim=0)
+        y = tensor_y.detach().numpy()  # type: ignore
+        return y
+
+
+class BinaryClassifier(abc.ABC):
```
Comment (ronuchit) [inline src/torch_models.py:143]: since there's non-torch stuff now, do we need to rename this file again? x.x (and the corresponding test file)

Comment (ronuchit) [reply src/torch_models.py:143]: i never liked the name models.py though, too generic

File: src/torch_models.py [hunk 4]
```diff
@@ -1,153 +1,323 @@
...
-    @abc.abstractmethod
-    def predict(self, arr: Array) -> Array:
+    def predict(self, x: Array) -> Array:
         """Return a prediction for the given datapoint.
 
-        arr is single-dimensional.
+        x is single-dimensional.
         """
-        raise NotImplementedError("Override me")
+        assert x.shape == (self._x_dim, )
```
Comment (ronuchit) [inline src/torch_models.py:61]: even before this, assert fit was called by checking that `self._x_dim > 0 and self._y_dim > 0`? (similar for the classifier below)

File: src/torch_models.py [hunk 5]
```diff
@@ -1,153 +1,323 @@
...
 
-    @abc.abstractmethod
     def fit(self, X: Array, Y: Array) -> None:
-        """Train regressor on the given data.
+        """Train the regressor on the given data.
 
         X and Y are both two-dimensional.
         """
-        raise NotImplementedError("Override me")
+        num_data, self._x_dim = X.shape
```
Comment (ronuchit) [inline src/torch_models.py:47]: why would x_dim and y_dim ever be different?

File: src/torch_models.py [hunk 6]
```diff
@@ -1,153 +1,323 @@
...
+
+        X is two-dimensional, y is one-dimensional.
+        """
+        num_data, self._x_dim = X.shape
+        assert y.shape == (num_data, )
+        logging.info(f"Training {self.__class__.__name__} on {num_data} "
+                     "datapoints")
+        # If there is only one class in the data, then there's no point in
+        # learning, since any predictions other than that one class could
+        # only be strange generalization issues.
```
Comment (ronuchit) [inline src/torch_models.py:171]: remove "strange"

File: src/torch_models.py [hunk 7]
```diff
@@ -1,153 +1,323 @@
...
+        if self._do_single_class_prediction:
+            return self._predicted_single_class
+        # Normalize.
+        x = (x - self._input_shift) / self._input_scale
+        # Make prediction.
+        return self._classify(x)
+
+    @abc.abstractmethod
+    def _fit(self, X: Array, y: Array) -> None:
+        """Train the regressor on normalized data."""
```
Comment (ronuchit) [inline src/torch_models.py:203]: classifier*?

File: src/torch_models.py [hunk 8]
```diff
@@ -1,153 +1,323 @@
...
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _classify(self, x: Array) -> bool:
+        """Return a predicted class for the normalized input."""
+        raise NotImplementedError("Override me!")
 
-    def forward(self, inputs: Array) -> Tensor:
+
+class PyTorchClassifier(BinaryClassifier, nn.Module):
```
Comment (ronuchit) [inline src/torch_models.py:212]: call this PyTorchBinaryClassifier

File: src/torch_models.py [hunk 9]
```diff
@@ -602,3 +606,85 @@ def classifier(self, state: State, objects: Sequence[Object]) -> bool:
...
+                                    loss_fn: Callable[[Tensor, Tensor],
+                                                      Tensor],
+                                    optimizer: optim.Optimizer,
+                                    tensor_X: Tensor,
+                                    tensor_Y: Tensor,
+                                    max_iters: int,
+                                    print_every: int = 1000,
+                                    clip_gradients: bool = False,
+                                    clip_value: float = 5,
+                                    n_iter_no_change: int = 10000000) -> None:
```
Comment (ronuchit) [inline src/torch_models.py:651]: use `1e7`

File: src/torch_models.py [hunk 10]
```diff
@@ -602,3 +606,85 @@ def classifier(self, state: State, objects: Sequence[Object]) -> bool:
...
+        if loss.item() < best_loss:
+            best_loss = loss.item()
+            best_itr = itr
+            # Save this best model.
+            torch.save(model.state_dict(), model_name)
+        if itr % print_every == 0:
+            logging.info(f"Loss: {loss:.5f}, iter: {itr}/{max_iters}")
+        optimizer.zero_grad()
+        loss.backward()  # type: ignore
+        if clip_gradients:
```
Comment (ronuchit) [inline src/torch_models.py:674]: is this a new feature?

File: src/torch_models.py [hunk 11]
```diff
@@ -1,153 +1,323 @@
...
+        self._initialize_net()
+        # Create the loss function.
+        loss_fn = self._create_loss_fn()
+        # Create the optimizer.
+        optimizer = self._create_optimizer()
+        # Convert data to tensors.
+        tensor_X = torch.from_numpy(np.array(X, dtype=np.float32))
+        tensor_y = torch.from_numpy(np.array(y, dtype=np.float32))
+        # Run training.
+        _train_predictive_pytorch_model(
```
Comment (ronuchit) [inline src/torch_models.py:254]: why not just call this `_train_pytorch_model`?

File: src/torch_models.py [hunk 12]
```diff
@@ -1,153 +1,323 @@
...
+        y = tensor_y.detach().numpy()  # type: ignore
+        proba = y.item()
+        assert 0 <= proba <= 1
+        return proba
+
+    def _classify(self, x: Array) -> bool:
+        return self._forward_single_input_np(x) > 0.5
+
+    def predict_proba(self, denorm_x: Array) -> float:
+        """Get the predicted probability that the input classifies to 1.
```
Comment (ronuchit) [inline src/torch_models.py:279]: move this fn up, so that public methods come before private ones in general

File: src/torch_models.py [hunk 13]
```diff
@@ -1,153 +1,323 @@
...
+
+        The input is unnormalized.
```
Comment (ronuchit) [inline src/torch_models.py:281]: The input is NOT normalized.

File: src/torch_models.py [hunk 14]
```diff
@@ -1,153 +1,323 @@
...
-
-class ImplicitMLPRegressor(Regressor):
+    def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
+        return nn.MSELoss()
+
+
+class ImplicitMLPRegressor(PyTorchRegressor):
     """A regressor implemented via an "energy function".
 
     Currently the energy function is treated as a binary classifier, which is
```
Comment (ronuchit) [inline src/torch_models.py:323]: treated -> implemented

File: src/torch_models.py [hunk 15]
```diff
@@ -295,305 +502,102 @@ def predict_sample(self, x: Array, rng: np.random.Generator) -> Array:
...
-        self.load_state_dict(torch.load(model_name))  # type: ignore
-        os.remove(model_name)
-        self.eval()  # switch to eval mode
-        yhat = self(X)
-        loss = loss_fn(yhat, y)
-        logging.info(f"Loaded best model with loss: {loss:.5f}")
-
-
-class MLPClassifierEnsemble(Classifier):
+class MLPClassifierEnsemble(BinaryClassifier):
```
Comment (ronuchit) [inline src/torch_models.py:559]: is it on purpose that this inherits from the base class, and not pytorchclassifier?

File: src/torch_models.py [hunk 16]
```diff
@@ -295,305 +502,102 @@ def predict_sample(self, x: Array, rng: np.random.Generator) -> Array:
...
 
     def classify(self, x: Array) -> bool:
+        # Each member maintains its own normalizers.
         avg = np.mean(self.predict_member_probas(x))
         classification = avg > 0.5
         assert classification in [True, False]
         return classification
 
+    def _fit(self, X: Array, y: Array) -> None:
+        raise NotImplementedError("Not used.")
```
Comment (ronuchit) [inline src/torch_models.py:586]: maybe change this to `raise Exception` since it never makes sense to implement it?

File: src/torch_models.py [hunk 17]
```diff
@@ -295,305 +502,102 @@ def predict_sample(self, x: Array, rng: np.random.Generator) -> Array:
...
+
+    def _classify(self, x: Array) -> bool:
+        raise NotImplementedError("Not used.")
```
Comment (ronuchit) [inline src/torch_models.py:589]: maybe change this to raise Exception since it never makes sense to implement it?

**Review by ronuchit — APPROVED** [review]
great refactor!

File: src/settings.py (hunk 1 above)
Comment (tomsilver) [reply src/settings.py:173]: well in the place where it's used, it's a kwarg that defaults to 1. but you're right that technically this changes some (unused) functionality... gonna leave it though

File: src/torch_models.py (hunk 9 above)
Comment (tomsilver) [reply src/torch_models.py:651]: that's a float though

File: src/torch_models.py (hunk 10 above)
Comment (tomsilver) [reply src/torch_models.py:674]: no, we already had it for regressor training

File: src/torch_models.py (hunk 15 above)
Comment (tomsilver) [reply src/torch_models.py:559]: yeah

File: src/torch_models.py (hunk 5 above)
Comment (ronuchit) [reply src/torch_models.py:47]: nvm, got confused

File: src/torch_models.py (hunk 9 above)
Comment (ronuchit) [reply src/torch_models.py:651]: gotcha
